src/couch/src/couch_db.erl

```diff
@@ -369,8 +379,153 @@ get_full_doc_info(Db, Id) ->
+
+    FoldFun = fun(#doc{id = DocId, body = {Props}}, SeqAcc) ->
+        case DocId of
+            <<?LOCAL_DOC_PREFIX, "purge-", _/binary>> ->
+                ClientSeq = couch_util:get_value(<<"purge_seq">>, Props),
+                case ClientSeq of
+                    CS when is_integer(CS), CS >= PurgeSeq - PurgeInfosLimit ->
+                        {ok, SeqAcc};
+                    CS when is_integer(CS) ->
+                        case purge_client_exists(Db, DocId, Props) of
+                            true -> {ok, erlang:min(CS, SeqAcc)};
+                            false -> {ok, SeqAcc}
+                        end;
+                    _ ->
+                        % If there's a broken doc we have to keep every
+                        % purge info until the doc is fixed or removed.
+                        Fmt = "Invalid purge doc '~s' with purge_seq '~w'",
+                        couch_log:error(Fmt, [DocId, ClientSeq]),
+                        {ok, erlang:min(OldestPurgeSeq, SeqAcc)}
+                end;
+            _ ->
+                {stop, SeqAcc}
+        end
+    end,
+    InitMinSeq = PurgeSeq - PurgeInfosLimit,
+    Opts = [
+        {start_key, list_to_binary(?LOCAL_DOC_PREFIX ++ "purge-")}
+    ],
+    {ok, MinIdxSeq} = couch_db:fold_local_docs(Db, FoldFun, InitMinSeq, Opts),
+    FinalSeq = case MinIdxSeq < PurgeSeq - PurgeInfosLimit of
+        true -> MinIdxSeq;
+        false -> erlang:max(0, PurgeSeq - PurgeInfosLimit)
+    end,
+    % Log a warning if we've got a purge sequence exceeding the
+    % configured threshold.
+    if FinalSeq >= (PurgeSeq - PurgeInfosLimit) -> ok; true ->
+        Fmt = "The purge sequence for '~s' exceeds configured threshold",
+        couch_log:warning(Fmt, [couch_db:name(Db)])
+    end,
+    FinalSeq.
+
+
+purge_client_exists(DbName, DocId, Props) ->
+    % Warn about clients that have not updated their purge
+    % checkpoints in the last "index_lag_warn_seconds"
+    LagWindow = config:get_integer(
+            "purge", "index_lag_warn_seconds", 86400), % Default 24 hours
+
+    {Mega, Secs, _} = os:timestamp(),
+    NowSecs = Mega * 1000000 + Secs,
+    LagThreshold = NowSecs - LagWindow,
+
+    try
+        CheckFun = get_purge_client_fun(DocId, Props),
+        Exists = CheckFun(DbName, DocId, Props),
+        if not Exists -> ok; true ->
+            Updated = couch_util:get_value(<<"updated_on">>, Props),
+            if is_integer(Updated) and Updated > LagThreshold -> ok; true ->
+                Diff = NowSecs - Updated,
+                Fmt = "Purge checkpoint '~s' not updated in ~p seconds",
+                couch_log:error(Fmt, [DocId, Diff])
+            end
+        end,
+        Exists
+    catch _:_ ->
+        % If we fail to check for a client we have to assume that
+        % it exists.
+        true
+    end.
+
+
+get_purge_client_fun(DocId, Props) ->
+    M0 = couch_util:get_value(<<"verify_module">>, Props),
```

Comment:
What are implications on security? Users can write `_local/` documents. That means that they can craft the doc which has `verify_module` and `verify_function` keys. Which means they would be able to invoke any function which matches the expected signature. What if there is some dangerous function in otp or couchdb. The function which receives the proplist and modifies beam behavior would do. I didn't figured out the exact signature yet. But it seems the attack surface is big.

Reply:
Thanks @iilyak for bringing this up. I can understand that users can create `_local/` document on their own, and specify `verify_module` and `verify_function`. The specified `M:F/1` will be called during database compaction. I wonder how the customized logic for `M:F/1` is injected into couchdb. If users only specifies module name and function name, and failed to tamper with called logic, we are kind of safe in some extent. Of course, we can see comments from @davisp

Reply:
I don't think we are safe. Any function in OTP which doesn't validate it's input can be used to mount an attack. For example let's assume any of the following functions don't handle the case when they receive proplist:
- `zip:extract/1` --- zip bomb attack
- `os:cmd/1` --- execute any command on the host
- `erlang:throw/1` --- crash the current process with attacker provided reason.
- `erlang:halt/1` -- stop node

Reply:
I agreed that there is still potential risk even if only module name and function name are specified by users. Perhaps there are mature methodology to solve such issue. For me, I thought to use blacklist to check whether specified M:F/1 is attack-able. However, this may not be enumerable. As you suggested, we may consider whitelist or something like this.

Reply:
Good call Ilya. Should be easy enough to add a module whitelist via EPI no? Then its just a matter of moving our validation logic to an otherwise empty module so that its trivial to verify we're not calling anything possibly dangerous.

Reply:
Ooh, or alternatively, just make an EPI hook for the verification rather than having each purge handler from specifying which function in the doc itself. That seems even better than the whitelist.

Reply:
Ah, good catch Ilya. Sorry I missed that!

Reply:
I tried to solve this problem using commit 407b1dc. More test will be added.